runtime/common/src/routing.rs

```diff
@@ -0,0 +1,53 @@
+use cfg_traits::liquidity_pools::{MessageSender, RouterSupport};
+use cfg_types::domain_address::{Domain, DomainAddress};
+use frame_support::dispatch::DispatchResultWithPostInfo;
+use pallet_axelar_router::AxelarId;
+use sp_std::marker::PhantomData;
+
+/// Identification of the router where the message is sent and received
+/// RouterId is more specific than Domain, because RouterId also identify by
+/// where the message is sent/received
+pub enum RouterId {
+	/// The message must be sent/received by EVM using Axelar
+	Axelar(AxelarId),
+}
+
+impl From<RouterId> for Domain {
+	fn from(router_id: RouterId) -> Domain {
+		match router_id {
+			RouterId::Axelar(AxelarId::Evm(chain_id)) => Domain::EVM(chain_id),
+		}
+	}
+}
+
+impl RouterSupport<Domain> for RouterId {
+	fn for_domain(domain: Domain) -> Vec<Self> {
+		match domain {
+			Domain::EVM(chain_id) => vec![RouterId::Axelar(AxelarId::Evm(chain_id))],
+			Domain::Centrifuge => vec![],
+		}
+	}
+}
```

Comment:
I am not sure we can safely map any EVM domain to the Axelar router. AFAICT, we need to set up a storage for that in the future. For instance, there exists Gnosis chain which will not have a direct router to Centrifuge Chain but use the unimplemented message relaying in which we wrap forwarded LP messages [like this](https://centrifuge.hackmd.io/TPH6TWh7SJat_pT0Fyty7A?both#Centrifuge-to-EVM). We want to include message relaying into LPV2.

______

In the following, I am thinking out loud how to extend your design for message relaying.

We need to be able to determine whether for a given domain, a forwarder is set. Though I think your design is fully compatible if we just add a `StorageMap<Domain, Path<Domain>>` where

```rust
	enum Path<Domain> {
		Domain(Domain),
		Forwarder {
			forwarding_domain: Domain,
			forwarding_contract: H160
		}
	}

	fn wrap_into_forwarded_msg(message: Vec<u8>, source_chain: Domain, forwarding_contract: H160) -> Result<Vec<u8>, DispatchError>;

```

and extend the `Middleware` of `pallets-axelar-router` to use `Path` instead of `AxelarId`

```diff
	impl<T: Config> MessageSender for Pallet<T> {
-		type Middleware = AxelarId;
+		type Middleware = Path<Domain>;
		type Origin = DomainAddress;

		fn send(
-			axelar_id: Path<Domain>,
+			path: Path<Domain>,
			origin: Self::Origin,
			message: Vec<u8>,
		) -> DispatchResultWithPostInfo {
+			let axelar_ids = match path {
				Domain(domain) => domain.for_domain(),
				Forwarder { forwarding_domain, .. } => forwarding_domain.for_domain()
			};
+			// NOTE: Not sure how to commit to an id if vec length > 1, for simplicity here taken first element
+			let axelar_id = axelar_ids.first().ok_or(Error::<T>::PathNotFound)?;

			let chain_name =
				ChainNameById::<T>::get(axelar_id).ok_or(Error::<T>::DomainNotFound)?;
			let config = Configuration::<T>::get(&chain_name).ok_or(Error::<T>::RouterNotFound)?;

			match config.domain {
				DomainConfig::Evm(evm_config) => {
					let sender_evm_address = H160::from_slice(&origin.address());

					let message = wrap_into_axelar_msg(
						message,
						chain_name.into_inner(),
						config.liquidity_pools_contract_address,
					)
					.map_err(DispatchError::Other)?;

+					let message = match path {
						Forwarder { forwarding_domain, forwarding_contract } => wrap_into_forwarded_msg(message, Domain::Centrifuge, forwarding_contract),
						_ => Ok(message),
					}?;

					T::Transactor::call(
						sender_evm_address,
						evm_config.target_contract_address,
						message.as_slice(),
						evm_config.fee_values.value,
						evm_config.fee_values.gas_price,
						evm_config.fee_values.gas_limit,
					)
				}
			}
		}
	}
```

Do you think this draft is somewhat reasonable based on your design proposal?

Reply:
Trying to understand this topic.

If what you need is, for a domain, knowing if there is a forwarded already set; why not retrieve that information in `RouterDomain::for_domain()`.

I think I get the point in the code you've shared but why the forwarder action needs to be in `pallet-axelar-precompile`. If we end up with 5 routers similars to the axelar one, we would like that section in a common part, isn't it? 

Why not implement this in `RouterDispatch`? if you need to store the map, maybe `RouterDispatch` can be a super simple pallet with just that map of forwarding info. That entity is in charge of determining the router, which seems similar to what you need.

Reply:
> I think I get the point in the code you've shared but why the forwarder action needs to be in `pallet-axelar-precompile`. If we end up with 5 routers similars to the axelar one, we would like that section in a common part, isn't it?
> 
> Why not implement this in `RouterDispatch`? if you need to store the map, maybe `RouterDispatch` can be a super simple pallet with just that map of forwarding info. That entity is in charge of determining the router, which seems similar to what you need.

Yeah that makes perfectly sense. Seems like I can very well align that feature with the new design then! I hope I can start implementing that tomorrow based on either this branch or #1948

Reply:
Super good news!